src/realm/array.hpp

```diff
@@ -1570,152 +1558,81 @@ template <bool gt, size_t width, class Callback>
 bool Array::find_gtlt(int64_t v, uint64_t chunk, QueryStateBase* state, size_t baseindex, Callback callback) const
 {
     // Find items in 'chunk' that are greater (if gt == true) or smaller (if gt == false) than 'v'. Fixme, __forceinline can make it crash in vS2010 - find out why
-    if (width == 1) {
-        for (size_t t = 0; t < 64; t++) {
-            if (gt ? static_cast<int64_t>(chunk & 0x1) > v : static_cast<int64_t>(chunk & 0x1) < v) {if (!find_action( t + baseindex, static_cast<int64_t>(chunk & 0x1), state, callback)) return false;}
+    if constexpr (width == 1) {
+        for (size_t i = 0; i < 64; ++i) {
+            int64_t v2 = static_cast<int64_t>(chunk & 0x1);
+            if (gt ? v2 > v : v2 < v) {
+                if (!find_action(i + baseindex, v2, state, callback)) {
+                    return false;
+                }
+            }
             chunk >>= 1;
         }
     }
-    else if (width == 2) {
-        // Alot (50% +) faster than loop/compiler-unrolled loop
```

Comment:
This change is without a doubt nicer to read and maintain. I'm curious if it this comment is still true though? I wonder if compilers got smarter about unrolling loops since it was written. My intuition is that most of the size savings discovered here are from the addition of `constexpr if` multiplied by each template dimension. Either way, it is not important if the performance we measure is not that much affected.

Reply:
I think the `if constexpr` didn't actually change anything (the compilers were smart enough to eliminate dead branches before that).

As for the loop unrolling, I could not see any difference in performance. I haven't checked in the generated code whether the loops get unrolled by the compiler, but if they don't, it's also possible that CPUs became smarter about branch prediction. :-) ARM64 CPUs these days are pretty good.